cluster-autoscaler/simulator/node_info_utils.go

```diff
@@ -0,0 +1,153 @@
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/framework"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/daemonset"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/labels"
+	pod_util "k8s.io/autoscaler/cluster-autoscaler/utils/pod"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/taints"
+)
+
+type nodeGroupTemplateNodeInfoGetter interface {
+	Id() string
+	TemplateNodeInfo() (*framework.NodeInfo, error)
+}
+
+// TemplateNodeInfoFromNodeGroupTemplate returns a template NodeInfo object based on NodeGroup.TemplateNodeInfo(). The template is sanitized, and only
+// contains the pods that should appear on a new Node from the same node group (e.g. DaemonSet pods).
+func TemplateNodeInfoFromNodeGroupTemplate(nodeGroup nodeGroupTemplateNodeInfoGetter, daemonsets []*appsv1.DaemonSet, taintConfig taints.TaintConfig) (*framework.NodeInfo, errors.AutoscalerError) {
+	id := nodeGroup.Id()
+	baseNodeInfo, err := nodeGroup.TemplateNodeInfo()
+	if err != nil {
+		return nil, errors.ToAutoscalerError(errors.CloudProviderError, err)
+	}
+	labels.UpdateDeprecatedLabels(baseNodeInfo.Node().ObjectMeta.Labels)
+
+	return TemplateNodeInfoFromExampleNodeInfo(baseNodeInfo, id, daemonsets, true, taintConfig)
+}
+
+// TemplateNodeInfoFromExampleNodeInfo returns a template NodeInfo object based on a real example NodeInfo from the cluster. The template is sanitized, and only
+// contains the pods that should appear on a new Node from the same node group (e.g. DaemonSet pods).
+func TemplateNodeInfoFromExampleNodeInfo(example *framework.NodeInfo, nodeGroupId string, daemonsets []*appsv1.DaemonSet, forceDaemonSets bool, taintConfig taints.TaintConfig) (*framework.NodeInfo, errors.AutoscalerError) {
+	randSuffix := fmt.Sprintf("%d", rand.Int63())
+	newNodeNameBase := fmt.Sprintf("template-node-for-%s", nodeGroupId)
+
+	// We need to sanitize the example before determining the DS pods, since taints are checked there, and
+	// we might need to filter some out during sanitization.
+	sanitizedExample := sanitizeNodeInfo(example, newNodeNameBase, randSuffix, &taintConfig)
+	expectedPods, err := podsExpectedOnFreshNode(sanitizedExample, daemonsets, forceDaemonSets, randSuffix)
+	if err != nil {
+		return nil, err
+	}
+	// No need to sanitize the expected pods again - they either come from sanitizedExample and were sanitized above,
+	// or were added by podsExpectedOnFreshNode and sanitized there.
+	return framework.NewNodeInfo(sanitizedExample.Node(), nil, expectedPods...), nil
+}
+
+// FreshNodeInfoFromTemplateNodeInfo duplicates the provided template NodeInfo, returning a fresh NodeInfo that can be injected into the cluster snapshot.
+// The NodeInfo is sanitized (names, UIDs are changed, etc.), so that it can be injected along other copies created from the same template.
+func FreshNodeInfoFromTemplateNodeInfo(template *framework.NodeInfo, suffix string) *framework.NodeInfo {
```

Comment:
I agree that while having `nodeInfo.SanitizedDeepCopy` would be nice, it is better to keep it in the same place instead of splitting the sanitization into multiple places. LGTM.